### ISSUE TITLE
False positive error "return-value" when subclass doesn't override classmethod

### ISSUE THREAD

**Issue by jamesbraza** [python/mypy#10921]
**Bug Report**

1. In a parent class, there is a `classmethod` used as a factory
2. The child class doesn't
    - Override that factory `classmethod`
    - Change its `__init__` signature.
3. In a function that uses the child class's factory `classmethod`, mypy errors, saying there's an incompatible return type

I think this is a false positive error message.

**To Reproduce**

```python
from __future__ import annotations

class Foo:
    @classmethod
    def make(cls) -> Foo:
        return Foo()

class DFoo(Foo):
    """Child class of Foo."""

def get_dfoo() -> DFoo:
    return DFoo.make()  # error: Incompatible return value type (got "Foo", expected "DFoo")  [return-value]
```

**Expected Behavior**

I think mypy should be able to realize that actually a `DFoo` will be made by the factory class method.  I believe this because we're saying `DFoo.make()`, not `Foo.make()`.

It seems mypy is interpreting the return type of the `classmethod(Foo.make)` as not being covariant, though I am not sure why it wouldn't interpret the `classmethod` as covariant.

I have looked through the issues for some time, and seen a lot of issues with _similar_ problems, but not exactly this one, so I am making a new issue.  I feel as if this issue is probably a duplicate, since it's pretty fundamental, but I haven't found what it would duplicate.

**Actual Behavior**

Here is the output of `mypy==0.910`:

```
mypy path/to/classmethod_return_covariance.py
path/to/classmethod_return_covariance.py: note: In function "get_dfoo":
path/to/classmethod_return_covariance.py:12:12: error: Incompatible return value type (got "Foo", expected "DFoo")  [return-value]
```

**Your Environment**

- Mypy version used: 0.910
- Mypy command-line flags: none notable
- Mypy configuration options from `mypy.ini` (and other config files): `show_error_codes = true`
- Python version used: Python 3.8.8
- Operating system and version: macOS Catalina, Version 10.15.7

**Comment by hauntsaninja** [comment]
Pretty sure that mypy is right, your example is incorrect code. `type(get_dfoo()) is Foo` is True.
If instead of `return Foo()` you did `return cls()` your code would work as you claim, but hopefully that shows you why mypy can't automatically infer things. For this sort of situation, you can use self types (https://mypy.readthedocs.io/en/stable/generics.html#generic-methods-and-generic-self)

**Comment by jamesbraza** [comment]
@hauntsaninja yes you are correct, it was supposed to say `return cls()` not `return Foo()` (and thank you for realizing that).  I agree, in the `return Foo()` case mypy was 100% correct.  I know this is a bad look, but I updated the code snippet in the OP to say `return cls()`.

I am aware of the generic self section of mypy.  However, I think in this case, mypy should be able to infer properly without generic self + covariant TypeVar stuff.  In other words, my issue was trying to say: I think mypy should not require adding the generic self + covariant TypeVar so it can infer updated typing from the child class, it should work out of the box.

Am I correct in interpreting that mypy can't properly inspect the return type of the `make` classmethod to be covariant, so it requires the generic self stuff?

**Comment by erictraut** [comment]
Your function `make` is annotated to return `Foo`. A type checker cannot override an annotation and assume that you meant something else. If your intent is for `make` to return an instance of `cls`, then you need to use generics to specify that intent.

```python
T = TypeVar("T")

class Foo:
    @classmethod
    def make(cls: Type[T]) -> T:
        return cls()
```

Incidentally, pyright is able to infer the generic return type if you leave `cls` and the return type unannotated, but mypy doesn't perform return type inference (it always assumes `Any` in this case), so return types for functions need to be explicitly annotated with mypy.

```python
class Foo:
    @classmethod
    def make(cls):
        return cls()

reveal_type(DFoo.make()) # pyright reveals "DFoo", mypy reveals "Any"
```

**Comment by jamesbraza** [comment]
Yes thank you @erictraut for adding more info! :)

I guess my point in this issue is that this error message:

> error: Incompatible return value type (got "Foo", expected "DFoo")

Is a false positive because I think mypy should be able to realize that `DFoo.make`'s call to `cls()` will create an instance of `DFoo`, not `Foo`.  I know the parent class's `classmethod` is type hinted with `Foo`, but I think a type checker should be able to realize that `cls()` will actually create an instance of the child class `DFoo`.

>  A type checker cannot override an annotation and assume that you meant something else.

I think in the case where `cls()` is used, the type hint should be assumed to be covariant.  What do you two think of that?

**Comment by erictraut** [comment]
If `make` is annotated to return `Foo`, then a type checker cannot assume that it returns `DFoo` in some cases. If a type can vary, then you should use a type variable. Generics are expressly designed for cases where a type can vary.

> I think ... the type hint should be assumed to be covariant

Variance is a concept that applies to type variables, not to concrete (non-generic) types, so I'm not sure what you mean by "the type hint should be assumed to be covariant".

**Comment by jamesbraza** [comment]
I am tracking what you say, and it makes sense a Generic is the tool to use here.

Thank you for fielding all my questions @erictraut! :)
